### PR TITLE
Add initial support for Google Pubsub

### PR DESCRIPTION
**Pull request by selmanj**
This change adds some (not all) support for Google Pubsub (https://cloud.google.com/pubsub/overview)

The basics are there (topic creation/deletion/publishing, subscription creation/deleting/pulling, message acknowledging). I also added a very simple mock implementation as well as an example.

### CODE REVIEW

**Comment by selmanj** [comment]
cc @erjohnso @Temikus @icco

File: lib/fog/google/models/pubsub/received_message.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        # @see https://cloud.google.com/pubsub/reference/rest/v1/projects.subscriptions/acknowledge
+        def acknowledge
+          requires :ack_id
+
+          service.acknowledge_subscription(@subscription_name, [ack_id])
+          nil
+        end
+
+        def reload
+          # Message is immutable - do nothing
```
Comment (plribeiro3000) [inline lib/fog/google/models/pubsub/received_message.rb]: Maybe you could add a Log warning instead?

File: lib/fog/google/models/pubsub/subscription.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
+require "base64"
+require "fog/core/model"
+require "fog/google/models/pubsub/received_message"
```
Comment (plribeiro3000) [inline lib/fog/google/models/pubsub/subscription.rb]: You might not need those extra requires here. Could you confirm you need them please?

**Comment by plribeiro3000** [comment]
@selmanj LGTM. I'm not that into `google` cloud api but mostly my concerns are about following the conventions so we will be sure of not adding regressions like fog/fog#3430.

You followed everything so we are good here. 👍 

But i will let the other guys take a look at the actual implementation because they will have more insights.

Thanks!

**Comment by icco** [comment]
Please add to the Readme as well.

File: lib/fog/google/models/pubsub/subscription.rb (hunk 2 above)
Comment (selmanj) [inline lib/fog/google/models/pubsub/subscription.rb]: I took out all three and the tests/examples passed.  I don't think they are required.

I'm curious though why we should remove them - my understanding was that require wouldn't load a file more than once (I'm still relatively new to ruby so I am likely missing something obvious)

File: lib/fog/google/models/pubsub/received_message.rb (hunk 1 above)
Comment (selmanj) [inline lib/fog/google/models/pubsub/received_message.rb]: Great idea!  Will do

File: lib/fog/google/models/pubsub/subscription.rb (hunk 2 above)
Comment (plribeiro3000) [inline lib/fog/google/models/pubsub/subscription.rb]: Yeah. The thing about it is that we don't need it anymore thus it is kind of bogus and a legacy of when this was still inside the main `fog` gem.

So mostly of the approaches and stuff we are doing is move to a solution that was needed inside the main `fog` gem to a solution more accurate to a standalone gem. e.g `autoload`.

**Comment by selmanj** [comment]
Updated README (hope it's worded alright)

**Comment by selmanj** [comment]
@plribeiro3000 Thanks for the review!

**Comment by selmanj** [comment]
I had to make a dependency change to enforce google-api-client on 0.8.6; unfortunately there is a bug in how google-api-client works on discovery documents that have a ':' in the path (which pubsub does)

See: https://github.com/google/google-api-ruby-client/commit/94a979b5126e6bb4e4a1779e512131ffe32525c1

**Comment by icco** [comment]
Oh jeezus. That's weird. Ok, well I guess when we upgrade to 0.9 we'll be saved from this. Could you please in the pubsub section mention that you have to have a version > 0.8.5 to use pubsub, because someone will fuck that up... I'm nearly positive.

**Comment by selmanj** [comment]
@icco Would it be better to cut a new major version with a required 0.8.6 dependency, or to continue allowing a range of dependencies but documenting in the pubsub area that previous versions won't work?

Personally I am leaning toward a new major version being less confusing.

**Comment by icco** [comment]
I'd do both, document why we're requiring the dependency, and do a version bump. But because of how fog dependencies work, people won't always notice. It's a production dependency that we can only put in dev because otherwise it makes all of Fog require it.

File: lib/fog/google/models/pubsub/subscriptions.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+          data = service.list_subscriptions.body["subscriptions"] || []
+          load(data)
+        end
+
+        # Retrieves a subscription by name
+        #
+        # @param subscription_name [String] name of subscription to retrieve
+        # @return [Fog::Google::Pubsub::Topic] topic found, or nil if not found
+        def get(subscription_name)
+          if subscription = service.get_subscription(subscription_name).body
```
Comment (Temikus) [inline lib/fog/google/models/pubsub/subscriptions.rb]: Conditional assignment. This is confusing and will always get called if `service.get_subscription(subscription_name).body` doesn't return an exception.

File: lib/fog/google/models/pubsub/topics.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+          data = service.list_topics.body["topics"] || []
+          load(data)
+        end
+
+        # Retrieves a topic by name
+        #
+        # @param topic_name [String] name of topic to retrieve
+        # @return [Fog::Google::Pubsub::Topic] topic found, or nil if not found
+        def get(topic_name)
+          if topic = service.get_topic(topic_name).body
```
Comment (Temikus) [inline lib/fog/google/models/pubsub/topics.rb]: Ditto for conditional assignment.

**Comment by Temikus** [comment]
Thanks for this @selmanj! 

I have 2 minor comments on conditional assignment + please address @icco's comment on prod dependencies - otherwise LGTM

File: lib/fog/google/models/pubsub/subscriptions.rb (hunk 3 above)
Comment (selmanj) [inline lib/fog/google/models/pubsub/subscriptions.rb]: You're right, it's much cleaner without.

File: lib/fog/google/models/pubsub/topics.rb (hunk 4 above)
Comment (selmanj) [inline lib/fog/google/models/pubsub/topics.rb]: Done.

**Comment by selmanj** [comment]
@Icco I updated the README as instructed and bumped the major version

**Comment by icco** [comment]
Cool!

**Comment by erjohnso** [comment]
Woot, nice work @selmanj!

Thanks for the reviews @icco, @Temikus, @plribeiro3000! 💯

**Pull request merged**